src/simulators/density_matrix/densitymatrix_state.hpp

```diff
@@ -518,6 +524,113 @@ void State<densmat_t>::snapshot_pauli_expval(const Operations::Op &op,
                             expval, variance);
 }
 
+
+template <class densmat_t>
+void State<densmat_t>::snapshot_density_matrix(const Operations::Op &op,
+                                               ExperimentData &data) {
+  cmatrix_t reduced_state;
+
+  // Check if tracing over all qubits
+  if (op.qubits.empty()) {
+    reduced_state = cmatrix_t(1, 1);
+    reduced_state[0] = BaseState::qreg_.trace();
+  } else {
+
+    auto qubits_sorted = op.qubits;
+    std::sort(qubits_sorted.begin(), qubits_sorted.end());
+
+    if ((op.qubits.size() == BaseState::qreg_.num_qubits()) && (op.qubits == qubits_sorted)) {
+      reduced_state = std::move(BaseState::qreg_.matrix());
+    } else {
+      reduced_state = std::move(reduced_density_matrix(op.qubits, qubits_sorted));
+    }
+  }
+
+  data.add_average_snapshot("density_matrix",
+                            op.string_params[0],
+                            BaseState::creg_.memory_hex(),
+                            std::move(reduced_state),
```

Comment:
The data container has special functions for move semantics, so passing the std move here is needed to use that. Otherwise reduced_state will be copied (and it is potentially a very large matrix).

Reply:
My point is that we don't need move semantics here because there's no need for transfer ownership and using the overloaded `add_average_snapshot` method that passes `reduced_state` as a const reference would avoid the copy and will be even faster than casting to rvalue.